lib/streamlit/watcher/event_based_path_watcher.py

```diff
@@ -168,14 +173,30 @@ def watch_path(
                 self._folder_handlers[folder_path] = folder_handler
 
                 folder_handler.watch = self._observer.schedule(
-                    folder_handler, folder_path, recursive=True
+                    folder_handler, folder_path, recursive=watch_recursively
+                )
+            elif folder_handler and watch_recursively:
+                # We've already created a _FolderEventHandler for this folder, which
+                # means that we must have previously watched a specific *file* in it (see
+                # the docstring of _FolderEventHandler if this sounds nonsensical). Now,
+                # we want to watch the folder itself. This requires that we reinstall our
+                # watcher recursively. Note that doing so doesn't mess up watchers we
+                # may have on any files.
+
+                # Sometimes watchdog's FileSystemEventHandler does not have
+                # a .watch property. It's unclear why -- may be due to a
+                # race condition.
+                if hasattr(folder_handler, "watch"):
+                    self._observer.unschedule(folder_handler.watch)
```

Comment:
This was copied from some code below in stop_watching_path. I didn't look into root-causing/fixing whatever may be causing this myself, but it doesn't seem particularly important to do so right now